### DISCUSSION TITLE
Midi Clock/BPM Master

### DISCUSSION THREAD

**Discussion by poncratias** [shanteacontrols/OpenDeck#164] [Ideas]
Hi, It would be great to be able to have Midi Clock Master Functions in OpenDeck.

In General i want to build a device that can give out a stable,jitterfree Midi Clock signal  or BPM to other gear like Synthesizers,drummachines, etc.
This is especially usefull if you want to synchronize multiple midi instruments to the same clock or bpm.
With the hype around synths today, a lot of people jam together or have multiple synths on their own, and one big problem is always to sync everything right on time.

It would be great if it is possible to build a device that can have:

-An encoder to set a BPM number
-Tap Tempo Button, tapping that button to a beat calculates the bpm, (lets say it takes the median time distance of the last 4 button hits, sees it as 4/4 beat, gives out the corresponding bpm)
-a start/stop button (sends midi start/stop signal)
-a reset button (restarts the cycle on all devices with the start of the next bar, usefull because in a longer jam sometimes instruments tend to drif off and out of sync after a while)
-a screen to show the bpm number (can be a simple 7-digit display)
-send clock signal to midi output , aswell as to a CV output (to sync modular synths etc)

there are similar things existing (but really rare and expensive, actually there are only 2 ):
https://www.e-rm.de/midiclock/ (this one misses a tap-tempo functionality)
https://www.kickstarter.com/projects/midronome/midronome-effortless-midi-synchronization (this one is cancelled!)

so you see, there is really market hole for something like this.

there are some arduino codes on github for midi clock, but sadly i lack the skills on getting that together, thats why i use OpenDeck;)

https://github.com/midilab/uClock

https://github.com/ElliotGarbus/MidiClockGenerator

https://forum.arduino.cc/t/tap-tempo-button-code/642650/5

https://create.arduino.cc/projecthub/victorman/arduino-metronome-tap-to-set-tempo-display-back-bpms-c13d46

https://github.com/dxinteractive/ArduinoTapTempo

Thx for all your work !

**Comment by poncratias** [comment]
also this looks really interesting functionalitywise, maybe it is possible to implement this code too
 https://midisizer.com/midigal/midiclk/

**Comment by AlpesMachines** [comment]
http://www.midibox.org/dokuwiki/doku.php?id=midimasterclock

**Reply by poncratias** [reply]
looks interesting, but also not finished as what i read?
also as i am already struggling with arduino to get it done, is the reason why i wanted to see if its would be possible with OpenDeck, hence my suggestion here;)

**Comment by paradajz** [comment]
This is possible, yeah. Do you have any recommendations for software I could use to verify clock drift / accuracy?

**Reply by poncratias** [reply]
woohoo!

not really, but i found this , where it gets talked about https://midisizer.com/midigal/midiclk/ 
essentially measuring it with a frequency meter 

which also opens an interesting point: there is always the problem of syncing midi clock coming from a DAW because of clock inaccuracy, and they seem to have found a way to code it so that this jitter gets erased

**Comment by paradajz** [comment]
Well, I mean, if I implement this, how will you determine whether there's unacceptable jitter? By ear or you have other way of determination / measurement?

**Reply by gh0stless** [reply]
The tool MIDI-OX shows the received midi-clock on the fraction.

**Comment by paradajz** [comment]
If anyone wants to test this, let me know which board are you using.

**Reply by gh0stless** [reply]
Unfortunately, I only have V.1 mainboards so far, but I plan to build V.2 hardware boards with which I would definitely want to use it.

**Reply by paradajz** [reply]
Consider supporting the project with buying the official board: https://www.tindie.com/products/paradajz/opendeck-diy-midi-platform-v3/

_Slightly off-topic, but a very-slight variation of v2 boards will also be available on Tindie in a month or two._

**Reply by gh0stless** [reply]
It would be important to me that the form factor of the board remains the same.

**Reply by poncratias** [reply]
Hey, I got the 2.0.1 Board, happy to test it out!

**Reply by basst22778** [reply]
i can help testing on blackpill401cc

**Comment by paradajz** [comment]
[opendeck2.zip](https://github.com/shanteacontrols/OpenDeck/files/9377158/opendeck2.zip)
[blackpill401cc.zip](https://github.com/shanteacontrols/OpenDeck/files/9395620/blackpill401cc.zip)

Here's the initial MIDI clock implementation (attached firmware update file for OpenDeck v2 board). The clock is sent out on DIN MIDI only for now. To start it, enable DIN MIDI, enable sending of MIDI clock, assign one of the buttons to Real Time Clock message type and then press the button.

Edit:

@basst22778 I've added the sysex binary for blackpill401cc

**Reply by gh0stless** [reply]
In any case I will buy a V.2 or V.3 hardware, only the money is tight at the moment.

**Reply by gh0stless** [reply]
I now have OpenDeck V.3.0.1 hardware at the start, would it be possible to recompile for it? Thank you Andrew

**Reply by gh0stless** [reply]
for the Black Pill Variant.

**Reply by paradajz** [reply]
[opendeck3b.zip](https://github.com/shanteacontrols/OpenDeck/files/10011969/opendeck3b.zip)

Here you go

**Reply by gh0stless** [reply]
During the first tests I didn't get anything to work. There are no continuous MIDI clock signals, Button->Inc BPM crashes the controller. Button->Dec BPM cannot be configured, neither can Encoder->BPM. (will not be saved)
There should be a global setting: enable/disable BPM Generator. The generator must always be running and is then controlled the reciver by using the START/STOP/Continoue commands.

**Reply by gh0stless** [reply]
Ah OK, I have to switch DIN-MIDI on first!

**Reply by gh0stless** [reply]
Hello Igor,

The clock generator works with MIDI sockets. Button Dec works too. Not Button-Inc, not even Encoder-BPM. Good work so far, thank you. Will the final version also support USB-MIDI? That would be very important to me.

**Reply by paradajz** [reply]
> There should be a global setting: enable/disable BPM Generator.

It already exists?

![image](https://user-images.githubusercontent.com/2544094/202873143-b141e160-6207-4575-a24b-06d93c2760cf.png)

Additionally, to actually start sending MIDI clock, "RealTime Start" and "RealTime Stop" message type should be assigned to buttons.

> Ah OK, I have to switch DIN-MIDI on first!

Right. I've removed the ability to change BPM while DIN MIDI is disabled. This was the reason controller was crashing.

> Button Dec works too.
 Not Button-Inc, not even Encoder-BPM

This was unrelated issue which is now fixed on master branch. I'm attaching the latest version for you to try out.

> Will the final version also support USB-MIDI?

I don't know. I need to check some low-level details first. For now DIN only.

[opendeck3b.zip](https://github.com/shanteacontrols/OpenDeck/files/10048010/opendeck3b.zip)

**Reply by gh0stless** [reply]
The clock signals come and can be controlled. But I can't seem to use it yet because the start/stop/continue commands aren't output on midi-din?

**Reply by paradajz** [reply]
I'm not sure I understand.

**Reply by gh0stless** [reply]
I assigned Button>Real Time Start and Button->Realtime Stop; but not recived this commands in my MIDI Monitor (MIDI OX).

**Reply by paradajz** [reply]
Ah, right. Those commands only start and stop sending of clock, if MIDI clock is enabled. Wasn't sure what to do in that case. I presume you want both functions, right (starting of clock *and* sending the start/stop MIDI message)?

**Reply by paradajz** [reply]
Current implementation:

![image](https://user-images.githubusercontent.com/2544094/202902675-81e93563-db1a-44de-b517-b833279ab1ce.png)

**Reply by gh0stless** [reply]
In my opinion this is completely independent. The generator just always sends clock signals. And the start/stop/continue messages can come at any time in between.

**Reply by paradajz** [reply]
Can you please describe your use case in more detail? I don't think we're on the same page here.

**Reply by gh0stless** [reply]
I think all you have to do is copy what is in the else branch to the if branch as well so that the messages are sent. continue works by the way.

**Reply by paradajz** [reply]
I was interested in what your use case is. Why does the instrument/device/whatever you're sending the clock to care about those messages?

**Reply by gh0stless** [reply]
These 3 messages synchronize sequencers. In the MIDI standard it is really the case that the clock generator does not have to stop when stop is sent. The MIDI Clock Generator should always be running. With the 3 commands you then control the device/sequencer. I am using an old software that I would like to control: "ReBirth 338". Apparently the implementation here is that it waits for a START message before synchronizing. Currently START and STOP are not sent on MIDI-DIN. Continue is sent.

**Comment by paradajz** [comment]
Thanks for the explanation. I'm ignorant on these things since it's something I don't use. Here's the updated firmware with the start/stop messages now being sent.

[opendeck3b.zip](https://github.com/shanteacontrols/OpenDeck/files/10049773/opendeck3b.zip)

**Reply by gh0stless** [reply]
I think an external clock generator is also interesting for performers. A function that has already been suggested would be interesting later on to "tap" the tempo with a button. Thank you for your efforts!

**Reply by gh0stless** [reply]
There is another problem with this new version: The MIDI clock generator stops generating MIDI clock messages with STOP. As said, the generator should ALWAYS be running and sending out a steady stream of MIDI clock signals. For me it is now the case that the continue function is of course crazy.

**Reply by paradajz** [reply]
Right. The clock is now sent out constantly, and the start/stop messages won't cause the clock to start/stop. If the "send MIDI clock" option is enabled, the clock is always sent. Only when you disable it will the clock stop.

[opendeck3b.zip](https://github.com/shanteacontrols/OpenDeck/files/10049981/opendeck3b.zip)

**Reply by gh0stless** [reply]
I can confirm: It now works as it should. Great. Thank you very much.

**Comment by gh0stless** [comment]
Maybe a Option to displaying the BPM and the last sended command (START/STOP/CONTINIUE) on I²C display?

**Comment by gh0stless** [comment]
Hi Igor, I was finally able to buy an original M board from you, thanks for the super fast delivery. Now I have everything together to continue working on my project. Can you compile a firmware for me with the BPM option for the M board? Thanks and Greetings.

**Comment by paradajz** [comment]
Here it is. Short circuit the 2-pin header named RPB before turning on the board and verify that a new disk drive RP2 appeared. Once it did, remove the short circuit and drag the uf2 file onto it from the archive.

Keep in mind that this branch is now over a year old and hasn't been tested in a very long time.

[opendeck22.zip](https://github.com/shanteacontrols/OpenDeck/files/12908874/opendeck22.zip)

**Comment by paradajz** [comment]
The feature is now available on the master branch. It's not supported on AVR MCUs (ATmega2560 and AT90USB1286) but it will work everywhere else. I encourage anyone interested in this to test it out.

**Comment by paradajz** [comment]
Feature released in v7.6.0
https://github.com/shanteacontrols/OpenDeck/releases/tag/v7.6.0

**Comment by mrbensinclair** [comment]
Hi, resurrecting this thread to see if it would be possible to add tap tempo? This would be a button message that calculates bpm from 2 or more button presses, and would be useful for tempo synching several delay and modulation pedals. Tap the button 2 or more times (eg counting 4 beats) and the tempo would be calculated using the average time between the taps.
